lib/rucio/core/transfer.py

```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+        if any(transfer['file_metadata']['src_type'] == 'TAPE' for transfer in path[1:]):
+            continue
+        filtered_candidate_paths.append(path)
+    candidate_paths = filtered_candidate_paths
+
+    # Discard multihop transfers which contain other candidate as part of itself For example:
+    # if A->B->C and B->C are both candidates, discard A->B->C because it includes B->C. Doing B->C is enough.
+    source_rses = {path[0]['file_metadata']['src_rse_id'] for path in candidate_paths}
+    filtered_candidate_paths = []
+    for path in candidate_paths:
+        if any(transfer['file_metadata']['src_rse_id'] in source_rses for transfer in path[1:]):
+            continue
+        filtered_candidate_paths.append(path)
+    candidate_paths = filtered_candidate_paths
+
+    def __transfer_order_key(transfer_path):
+        # higher source_ranking first,
+        # on equal source_ranking, prefer DISK over TAPE
+        # on equal type, prefer lower distance_ranking
+        # on equal distance, prefer single hop
+        return (
+            - transfer_path[-1]['sources'][0][3],
+            transfer_path[0]['file_metadata']['src_type'].lower(),  # rely on the fact that "disk" < "tape" in string order
+            transfer_path[-1]['sources'][0][4],
+            len(transfer_path) > 1,  # rely on the fact that False < True
+        )
+
+    best_candidate = min(candidate_paths, key=__transfer_order_key)
+
+    if best_candidate[0]['file_metadata']['src_type'] == 'TAPE':
+        # multiple Tape source replicas are not allowed in FTS3.
+        # Don't add any additional sources
+        return best_candidate
+
+    if len(best_candidate) > 1:
+        # only one source is allowed for multihop transfers to avoid spawning
+        # intermediate transfers all over the place
+        return best_candidate
+
+    for candidate in candidate_paths:
+        if candidate is best_candidate:
+            continue
+
+        if best_candidate[0]['file_metadata']['src_type'] != candidate[0]['file_metadata']['src_type']:
+            # Tape and Disk sources must not be used at the same time.
+            continue
+
+        if len(candidate) > 1:
+            # do not add multihop sources
+            continue
+
+        if not set(best_candidate[0]['schemes']).intersection(candidate[0]['schemes']):
+            # new sources are not compatible with old ones
+            continue
+
+        # We are in the case of single-hop DISK candidates. Merge the sources.
+        best_candidate[0]['sources'].extend(candidate[0]['sources'])
+        # If one source has force ipv4, force it for all sources
+        if best_candidate[0].get('use_ipv4') or candidate[0].get('use_ipv4'):
+            best_candidate[0]['use_ipv4'] = True
+
+    return best_candidate
+
+
 @transactional_session
 def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0, limit=None, activity=None, older_than=None, rses=None, schemes=None,
```

Comment:
Add a general description here about the further workflow. Nothing huge, just to ilustrate the steps being done to build the chosen transfers.

Reply:
done. added a short workflow description at the end of the docstring